broadinstitute/single_cell_portal_core
# Array label splitting in cluster plots (SCP-4269)

As shown in demo, this enables array-based columns to be split out into individual labels, and then multiple points per bead are plotted.  The backend change is a new field in `ClusterFileInfo` that is a hash of annotation names to boolean values of whether the annotation should be split by default.

The frontend reads that default to determine whether to initially split the arrays, and then also displays a control above the legend for toggling the display.  the 'jitter' will plot points in a 3x3 grid around the center point, and can therefore handle up to 9 distinct points per cell/bead.  
![image](https://user-images.githubusercontent.com/2800795/164498027-d0222e69-0f12-48e2-9ea4-49e3cbb369da.png)

For performance, this has almost no impact on regular plots.  We do add an iteration over all labels to check for '|' characters to determine whether to show the split control (which takes ~0.5sec per 1MM labels).  But that iteration happens after the Plot renders.  Rendering the split plot is determined by how many additional points are created (so, e.g. a 100K cluster split into 400K points will take about as long to render as a 400K cluster, minus server loading time)

This currently has no way for users to control the default setting.  Jonah seemed ambivalent about how he'd want it set, and he has so many clusters we'll probably want to update his study in the console anyway.

This also updates custom_colors to use the technique Emily developed for DownloadRequest of storing hashes with potentially forbidden key characters as Strings, rather than base64 encoding.  Storing them as strings is simpler (especially given nested hash/array objects) than base64 encoding keys, and also makes the in-db content more human readable.  If we have more fields like these in the future, we might want to consider making a custom Mongoid `hash_as_string` field type



TO TEST:
1.  Load a study with array-based labels.  You can run `SyntheticStudyPopulator.populate('array_mouse_eye')` for an example
2.  view the cluster
3.  select the array-based annotation ('cell_type_array') for the 'array_mouse_eye' study
4. Confirm the cluster plot appears normally, and with the link to 'Split array labels'
5. Click the 'Split array labels' link, confirm the labels are split into individual types.  You may have to zoom in/out to see the individual points.
